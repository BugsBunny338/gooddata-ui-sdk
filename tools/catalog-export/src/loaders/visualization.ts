// (C) 2007-2019 GoodData Corporation
import gooddata from "@gooddata/gd-bear-client";
import { ObjectMeta } from "../base/types";

/**
 * Loads information about insights defined in the project. Only descriptive information about
 * each insight is returned.
 *
 * @param projectId - project to get insights from
 * @return array of insight metadata
 */
export async function loadInsights(projectId: string): Promise<ObjectMeta[]> {
    const visualizations = await gooddata.md.getVisualizations(projectId);

    return visualizations.map((vis: any) => {
        return {
            identifier: vis.identifier,
            tags: vis.tags,
            title: vis.title,
        };
    });
}
