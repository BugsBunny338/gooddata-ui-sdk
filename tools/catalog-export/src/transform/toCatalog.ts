// (C) 2007-2019 GoodData Corporation
import { ProjectMetadata, Attribute } from "../base/types";
import { createUniqueName } from "./titles";
import { get, set, cloneDeep, findKey, forOwn, isEmpty } from "lodash";
/*
 * This transformation takes project metadata and creates an object that matches the required input of CatalogHelper.
 */

//
// Types - copied from CatalogHelper in sdk-ui; don't want to import them from there as that would mean they
//  be part of public API and i'm unsure whether that's a good idea
//

interface IIdentifierWithTags {
    identifier: string;
    tags: string;
}

interface IDisplayForms {
    [key: string]: IIdentifierWithTags;
}

interface IAttrItem {
    identifier: string;
    tags: string;
    displayForms: IDisplayForms;
    defaultDisplayForm: IIdentifierWithTags;
}

interface IAttrs {
    [key: string]: IAttrItem;
}

interface IDataSet {
    identifier: string;
    tags: string;
    attributes: IAttrs;
}

interface ICatalog {
    projectId: string;
    measures?: TitleToItemMap;
    visualizations: TitleToItemMap;
    attributes: IAttrs;
    dateDataSets: TitleToDataSet;
}

type TitleToDataSet = { [key: string]: IDataSet };
type TitleToItemMap = { [key: string]: IIdentifierWithTags };

//
// transformation functions
//

function createMeasures(projectMeta: ProjectMetadata): TitleToItemMap {
    const newMapping: TitleToItemMap = {};

    projectMeta.catalog.metrics.forEach(metric => {
        const { title, identifier, tags } = metric.metric.meta;
        const uniqueTitle = createUniqueName(title, newMapping);
        newMapping[uniqueTitle] = {
            identifier,
            tags,
        };
    });

    projectMeta.catalog.facts.forEach(fact => {
        const { title, identifier, tags } = fact.fact.meta;
        const uniqueTitle = createUniqueName(title, newMapping);
        newMapping[uniqueTitle] = { identifier, tags };
    });

    return newMapping;
}

function createAttributes(attributes: Attribute[]): IAttrs {
    const newAttrs: IAttrs = {};

    attributes.forEach(attr => {
        const uniqueAttrTitle = createUniqueName(attr.attribute.meta.title, newAttrs);
        const newDisplayForms: IDisplayForms = {};
        let firstDisplayForm: IIdentifierWithTags | undefined;

        attr.attribute.content.displayForms.forEach(df => {
            const { title, identifier, tags } = df.meta;
            const uniqueDfTitle = createUniqueName(title, newDisplayForms);
            const newDisplayForm: IIdentifierWithTags = {
                identifier,
                tags,
            };

            if (!firstDisplayForm) {
                firstDisplayForm = newDisplayForm;
            }

            newDisplayForms[uniqueDfTitle] = newDisplayForm;
        });

        if (firstDisplayForm) {
            const { identifier, tags } = attr.attribute.meta;
            const newAttr: IAttrItem = {
                identifier,
                tags,
                defaultDisplayForm: firstDisplayForm,
                displayForms: newDisplayForms,
            };

            newAttrs[uniqueAttrTitle] = newAttr;
        }
    });

    return newAttrs;
}

function createCatalogAttributes(projectMeta: ProjectMetadata): IAttrs {
    return createAttributes(projectMeta.catalog.attributes);
}

function createDateDatasets(projectMeta: ProjectMetadata): TitleToDataSet {
    const newDataSets: TitleToDataSet = {};

    projectMeta.dateDataSets.forEach(dd => {
        const { title, identifier, tags } = dd.dateDataSet.meta;
        const uniqueDsTitle = createUniqueName(title, newDataSets);
        const attributes: IAttrs = createAttributes(dd.dateDataSet.content.attributes);
        const newDataSet: IDataSet = {
            identifier,
            tags,
            attributes,
        };

        newDataSets[uniqueDsTitle] = newDataSet;
    });

    return newDataSets;
}

function createVisualizations(projectMeta: ProjectMetadata): TitleToItemMap {
    const newMapping: TitleToItemMap = {};

    projectMeta.insights.forEach(insight => {
        const { title, identifier, tags } = insight;
        const uniqueTitle = createUniqueName(title, newMapping);
        newMapping[uniqueTitle] = { identifier, tags };
    });

    return newMapping;
}

/**
 * Merges new and existing catalog by item identifiers. The logic is as follows:
 *
 * - If the item with same identifier is in the existing catalog, the title from the existing
 *   catalog SHOULD be retained
 *
 * - The title for existing item will be retained unless it conflicts with title for another
 *   an item in the new catalog. In that case item from existing catalog will have the title
 *   changed, sequence number will be added
 *
 * - Merging is done recursively if item is an object with predefined keys (df, attributes) =>
 *   this is thus currently triggered for date data sets and attributes.
 *
 * @param newCatalog - newly generated catalog
 * @param existingCatalog - existing catalog contents (may be empty)
 * @param mergePaths - top level properties to merge on
 * @param subItemKeys - second level properties to merge on recursively
 */
function mergeData(
    newCatalog: any,
    existingCatalog: any,
    mergePaths = ["visualizations", "measures", "attributes", "dateDataSets"],
    subItemKeys = ["displayForms", "attributes"],
): ICatalog {
    const result = cloneDeep(newCatalog);

    mergePaths.forEach(path => {
        set(result, path, {});
        const newItems = get(newCatalog, path);
        const existingItems = get(existingCatalog, path);

        forOwn(newItems, (newItem, newItemKey) => {
            const existingTitle = findKey(existingItems, item => item.identifier === newItem.identifier);
            const resolvedTitle = existingTitle ? existingTitle : newItemKey;
            const nonConflictingTitle = createUniqueName(resolvedTitle, get(result, path));

            if (resolvedTitle !== nonConflictingTitle) {
                console.warn("resolving duplicate key", resolvedTitle, "into", nonConflictingTitle); // eslint-disable-line no-console
            }

            let resultItem = newItem;
            if (existingTitle) {
                subItemKeys.forEach(subItemKey => {
                    if (resultItem.hasOwnProperty(subItemKey)) {
                        resultItem = mergeData(resultItem, existingItems[existingTitle], [subItemKey]);
                    }
                });
            }

            set(result, [path, nonConflictingTitle], resultItem);
        });
    });

    return result;
}

/**
 * Transforms project metadata into the sdk-ui catalog format. The catalog format can be serialized
 * and later ingested by the CatalogHelper in the sdk-ui.
 *
 * @param projectMeta - project metadata to work with
 * @param existingCatalog - existing catalog structure to merge with
 */
export function transformToCatalog(projectMeta: ProjectMetadata, existingCatalog?: any): any {
    const measures = createMeasures(projectMeta);
    const measuresProp = !isEmpty(measures) ? { measures } : {};
    const attributes = createCatalogAttributes(projectMeta);
    const dateDataSets = createDateDatasets(projectMeta);
    const visualizations = createVisualizations(projectMeta);

    const newCatalog: ICatalog = {
        projectId: projectMeta.projectId,
        ...measuresProp,
        attributes,
        visualizations,
        dateDataSets,
    };

    if (existingCatalog) {
        return mergeData(newCatalog, existingCatalog);
    }

    return newCatalog;
}
