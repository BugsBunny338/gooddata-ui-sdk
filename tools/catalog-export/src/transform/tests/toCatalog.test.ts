// (C) 2007-2019 GoodData Corporation
import { loadProjectMetadata } from "../../loaders/loadProjectMetadata";
import { transformToCatalog } from "../toCatalog";

jest.mock("@gooddata/gd-bear-client");

describe("transformToCatalog", () => {
    const projectMeta = loadProjectMetadata("test");

    it("creates new catalog", async () => {
        expect(transformToCatalog(await projectMeta)).toMatchSnapshot();
    });

    it("merges existing catalog", async () => {
        const existingCatalog = {
            projectId: "test",
            attributes: {
                "Renamed Channel": {
                    identifier: "attr.orders.channel",
                    tags: "",
                    displayForms: {
                        "Renamed Channel Df": {
                            identifier: "label.orders.channel",
                            tags: "",
                        },
                    },
                },
            },
            visualizations: {
                "My Renamed Total Sales By Location": {
                    identifier: "aaBN5UG3dXu4",
                    tags: "",
                },
                // this results in a merge conflict. the new catalog HAS item with
                // different identifier AND the same name
                "Total Sales by City": {
                    identifier: "abhJpedgcfU2",
                    tags: "",
                },
            },
        };

        expect(transformToCatalog(await projectMeta, existingCatalog)).toMatchSnapshot();
    });
});
