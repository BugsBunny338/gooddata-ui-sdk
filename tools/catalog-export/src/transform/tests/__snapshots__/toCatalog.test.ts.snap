// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformToCatalog creates new catalog 1`] = `
Object {
  "attributes": Object {
    "Channel": Object {
      "defaultDisplayForm": Object {
        "identifier": "label.orders.channel",
        "tags": "",
      },
      "displayForms": Object {
        "Channel": Object {
          "identifier": "label.orders.channel",
          "tags": "",
        },
      },
      "identifier": "attr.orders.channel",
      "tags": "",
    },
    "Employee Name": Object {
      "defaultDisplayForm": Object {
        "identifier": "label.employee.employeename",
        "tags": "",
      },
      "displayForms": Object {
        "Employee Name": Object {
          "identifier": "label.employee.employeename",
          "tags": "",
        },
        "Employee URL": Object {
          "identifier": "label.employee.employeename.employeeurl",
          "tags": "",
        },
      },
      "identifier": "attr.employee.employeename",
      "tags": "",
    },
  },
  "dashboards": Object {
    "KPIs": Object {
      "identifier": "afMA17GSbk31",
      "tags": "",
    },
    "Store management KPIs": Object {
      "identifier": "abBJlnxrfEWH",
      "tags": "",
    },
    "Untitled": Object {
      "identifier": "aahlrtUEgKkq",
      "tags": "",
    },
  },
  "dateDataSets": Object {
    "Date (Timeline)": Object {
      "attributes": Object {
        "Date (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.date.mmddyyyy",
            "tags": "",
          },
          "displayForms": Object {
            "Long (Mon, Jan 1, 2010) (Timeline)": Object {
              "identifier": "timeline.date.long",
              "tags": "",
            },
            "dd-mm-yyyy (Timeline)": Object {
              "identifier": "timeline.date.eddmmyyyy",
              "tags": "",
            },
            "dd/mm/yyyy (Timeline)": Object {
              "identifier": "timeline.date.ddmmyyyy",
              "tags": "",
            },
            "m/d/yy (no leading zeroes) (Timeline)": Object {
              "identifier": "timeline.date.mdyy",
              "tags": "",
            },
            "mm/dd/yyyy (Timeline)": Object {
              "identifier": "timeline.date.mmddyyyy",
              "tags": "",
            },
            "yyyy-mm-dd (Timeline)": Object {
              "identifier": "timeline.date.yyyymmdd",
              "tags": "",
            },
          },
          "identifier": "timeline.date",
          "tags": "date",
        },
        "Day of Month (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aca81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "default (Timeline)": Object {
              "identifier": "timeline.aca81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.day.in.month",
          "tags": "date day month",
        },
        "Day of Quarter (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.ab481lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "default (Timeline)": Object {
              "identifier": "timeline.ab481lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.day.in.quarter",
          "tags": "date day quarter",
        },
        "Day of Week (Mon-Sun) (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.abU81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Long (Monday) (Timeline)": Object {
              "identifier": "timeline.abY81lMifn6q",
              "tags": "",
            },
            "Number (1=Monday) (Timeline)": Object {
              "identifier": "timeline.abW81lMifn6q",
              "tags": "",
            },
            "Short (Mon) (Timeline)": Object {
              "identifier": "timeline.abU81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.day.in.euweek",
          "tags": "date day eu week",
        },
        "Day of Week (Sun-Sat) (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.abK81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Long (Sunday) (Timeline)": Object {
              "identifier": "timeline.abO81lMifn6q",
              "tags": "",
            },
            "Number (1=Sunday) (Timeline)": Object {
              "identifier": "timeline.abM81lMifn6q",
              "tags": "",
            },
            "Short (Sun) (Timeline)": Object {
              "identifier": "timeline.abK81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.day.in.week",
          "tags": "date day week",
        },
        "Day of Year (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.abE81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "default (Timeline)": Object {
              "identifier": "timeline.abE81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.day.in.year",
          "tags": "date day year",
        },
        "Month (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.abm81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Long (January) (Timeline)": Object {
              "identifier": "timeline.abs81lMifn6q",
              "tags": "",
            },
            "M/Q (M1/Q1) (Timeline)": Object {
              "identifier": "timeline.abo81lMifn6q",
              "tags": "",
            },
            "Number (M1) (Timeline)": Object {
              "identifier": "timeline.abq81lMifn6q",
              "tags": "",
            },
            "Short (Jan) (Timeline)": Object {
              "identifier": "timeline.abm81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.month.in.year",
          "tags": "date month year",
        },
        "Month of Quarter (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aby81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Number (Timeline)": Object {
              "identifier": "timeline.aby81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.month.in.quarter",
          "tags": "date month quarter",
        },
        "Month/Year (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.act81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Long (January 2010) (Timeline)": Object {
              "identifier": "timeline.acx81lMifn6q",
              "tags": "",
            },
            "Number (1/2010) (Timeline)": Object {
              "identifier": "timeline.acv81lMifn6q",
              "tags": "",
            },
            "Short (Jan 2010) (Timeline)": Object {
              "identifier": "timeline.act81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.month",
          "tags": "date month",
        },
        "Quarter (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aam81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "default (Timeline)": Object {
              "identifier": "timeline.aam81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.quarter.in.year",
          "tags": "date quarter year",
        },
        "Quarter/Year (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aci81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "US Short (Timeline)": Object {
              "identifier": "timeline.aci81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.quarter",
          "tags": "date quarter",
        },
        "Week (Mon-Sun) (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aba81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Number EU (Timeline)": Object {
              "identifier": "timeline.aba81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.euweek.in.year",
          "tags": "date eu week year",
        },
        "Week (Mon-Sun) of Qtr (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.abg81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Number EU (Timeline)": Object {
              "identifier": "timeline.abg81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.euweek.in.quarter",
          "tags": "date eu quarter week",
        },
        "Week (Mon-Sun)/Year (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aa281lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "From - To (Timeline)": Object {
              "identifier": "timeline.aaW81lMifn6q",
              "tags": "",
            },
            "Week #/Year (W1/2010) (Timeline)": Object {
              "identifier": "timeline.aa281lMifn6q",
              "tags": "",
            },
            "Week Starting (Timeline)": Object {
              "identifier": "timeline.aaY81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.euweek",
          "tags": "date eu week",
        },
        "Week (Sun-Sat) (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aaI81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Number US (Timeline)": Object {
              "identifier": "timeline.aaI81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.week.in.year",
          "tags": "date week year",
        },
        "Week (Sun-Sat) of Qtr (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aaO81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Number US (Timeline)": Object {
              "identifier": "timeline.aaO81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.week.in.quarter",
          "tags": "date quarter week",
        },
        "Week (Sun-Sat)/Year (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aaA81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "From - To (Timeline)": Object {
              "identifier": "timeline.aau81lMifn6q",
              "tags": "",
            },
            "Week #/Year (Cont.) (Timeline)": Object {
              "identifier": "timeline.aay81lMifn6q",
              "tags": "",
            },
            "Week #/Year (W1/2010) (Timeline)": Object {
              "identifier": "timeline.aaA81lMifn6q",
              "tags": "",
            },
            "Week Starting (Timeline)": Object {
              "identifier": "timeline.aaw81lMifn6q",
              "tags": "",
            },
            "Wk/Qtr/Year (Cont.) (Timeline)": Object {
              "identifier": "timeline.aaC81lMifn6q",
              "tags": "",
            },
            "Wk/Qtr/Year (Timeline)": Object {
              "identifier": "timeline.aas81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.week",
          "tags": "date week",
        },
        "Year (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aag81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Year (Timeline)": Object {
              "identifier": "timeline.aag81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.year",
          "tags": "date year",
        },
      },
      "identifier": "timeline.dataset.dt",
      "tags": "",
    },
  },
  "measures": Object {
    "$ Menu Item Sales": Object {
      "identifier": "fact.salesdetailfact.menuitemsales",
      "tags": "",
    },
    "Cost": Object {
      "identifier": "fact.restaurantcostsfact.cost",
      "tags": "",
    },
    "_Filter Last 4 Quarter": Object {
      "identifier": "aaIHiWZjfWNA",
      "tags": "hidden",
    },
    "_Filter Quarter": Object {
      "identifier": "aaiF4bffe4yn",
      "tags": "hidden",
    },
  },
  "projectId": "test",
  "visualizations": Object {
    "$ Total Sales by Location": Object {
      "identifier": "aaBN5UG3dXu4",
      "tags": "",
    },
    "Labor Costs vs Scheduled Costs": Object {
      "identifier": "abhJpedgcfU2",
      "tags": "",
    },
    "Total Sales by City": Object {
      "identifier": "acPWDTL2bJeX",
      "tags": "",
    },
  },
}
`;

exports[`transformToCatalog merges existing catalog 1`] = `
Object {
  "attributes": Object {
    "Employee Name": Object {
      "defaultDisplayForm": Object {
        "identifier": "label.employee.employeename",
        "tags": "",
      },
      "displayForms": Object {
        "Employee Name": Object {
          "identifier": "label.employee.employeename",
          "tags": "",
        },
        "Employee URL": Object {
          "identifier": "label.employee.employeename.employeeurl",
          "tags": "",
        },
      },
      "identifier": "attr.employee.employeename",
      "tags": "",
    },
    "Renamed Channel": Object {
      "defaultDisplayForm": Object {
        "identifier": "label.orders.channel",
        "tags": "",
      },
      "displayForms": Object {
        "Renamed Channel Df": Object {
          "identifier": "label.orders.channel",
          "tags": "",
        },
      },
      "identifier": "attr.orders.channel",
      "tags": "",
    },
  },
  "dashboards": Object {
    "KPIs": Object {
      "identifier": "afMA17GSbk31",
      "tags": "",
    },
    "Store management KPIs": Object {
      "identifier": "abBJlnxrfEWH",
      "tags": "",
    },
    "Untitled": Object {
      "identifier": "aahlrtUEgKkq",
      "tags": "",
    },
  },
  "dateDataSets": Object {
    "Date (Timeline)": Object {
      "attributes": Object {
        "Date (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.date.mmddyyyy",
            "tags": "",
          },
          "displayForms": Object {
            "Long (Mon, Jan 1, 2010) (Timeline)": Object {
              "identifier": "timeline.date.long",
              "tags": "",
            },
            "dd-mm-yyyy (Timeline)": Object {
              "identifier": "timeline.date.eddmmyyyy",
              "tags": "",
            },
            "dd/mm/yyyy (Timeline)": Object {
              "identifier": "timeline.date.ddmmyyyy",
              "tags": "",
            },
            "m/d/yy (no leading zeroes) (Timeline)": Object {
              "identifier": "timeline.date.mdyy",
              "tags": "",
            },
            "mm/dd/yyyy (Timeline)": Object {
              "identifier": "timeline.date.mmddyyyy",
              "tags": "",
            },
            "yyyy-mm-dd (Timeline)": Object {
              "identifier": "timeline.date.yyyymmdd",
              "tags": "",
            },
          },
          "identifier": "timeline.date",
          "tags": "date",
        },
        "Day of Month (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aca81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "default (Timeline)": Object {
              "identifier": "timeline.aca81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.day.in.month",
          "tags": "date day month",
        },
        "Day of Quarter (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.ab481lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "default (Timeline)": Object {
              "identifier": "timeline.ab481lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.day.in.quarter",
          "tags": "date day quarter",
        },
        "Day of Week (Mon-Sun) (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.abU81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Long (Monday) (Timeline)": Object {
              "identifier": "timeline.abY81lMifn6q",
              "tags": "",
            },
            "Number (1=Monday) (Timeline)": Object {
              "identifier": "timeline.abW81lMifn6q",
              "tags": "",
            },
            "Short (Mon) (Timeline)": Object {
              "identifier": "timeline.abU81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.day.in.euweek",
          "tags": "date day eu week",
        },
        "Day of Week (Sun-Sat) (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.abK81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Long (Sunday) (Timeline)": Object {
              "identifier": "timeline.abO81lMifn6q",
              "tags": "",
            },
            "Number (1=Sunday) (Timeline)": Object {
              "identifier": "timeline.abM81lMifn6q",
              "tags": "",
            },
            "Short (Sun) (Timeline)": Object {
              "identifier": "timeline.abK81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.day.in.week",
          "tags": "date day week",
        },
        "Day of Year (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.abE81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "default (Timeline)": Object {
              "identifier": "timeline.abE81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.day.in.year",
          "tags": "date day year",
        },
        "Month (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.abm81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Long (January) (Timeline)": Object {
              "identifier": "timeline.abs81lMifn6q",
              "tags": "",
            },
            "M/Q (M1/Q1) (Timeline)": Object {
              "identifier": "timeline.abo81lMifn6q",
              "tags": "",
            },
            "Number (M1) (Timeline)": Object {
              "identifier": "timeline.abq81lMifn6q",
              "tags": "",
            },
            "Short (Jan) (Timeline)": Object {
              "identifier": "timeline.abm81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.month.in.year",
          "tags": "date month year",
        },
        "Month of Quarter (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aby81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Number (Timeline)": Object {
              "identifier": "timeline.aby81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.month.in.quarter",
          "tags": "date month quarter",
        },
        "Month/Year (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.act81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Long (January 2010) (Timeline)": Object {
              "identifier": "timeline.acx81lMifn6q",
              "tags": "",
            },
            "Number (1/2010) (Timeline)": Object {
              "identifier": "timeline.acv81lMifn6q",
              "tags": "",
            },
            "Short (Jan 2010) (Timeline)": Object {
              "identifier": "timeline.act81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.month",
          "tags": "date month",
        },
        "Quarter (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aam81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "default (Timeline)": Object {
              "identifier": "timeline.aam81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.quarter.in.year",
          "tags": "date quarter year",
        },
        "Quarter/Year (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aci81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "US Short (Timeline)": Object {
              "identifier": "timeline.aci81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.quarter",
          "tags": "date quarter",
        },
        "Week (Mon-Sun) (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aba81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Number EU (Timeline)": Object {
              "identifier": "timeline.aba81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.euweek.in.year",
          "tags": "date eu week year",
        },
        "Week (Mon-Sun) of Qtr (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.abg81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Number EU (Timeline)": Object {
              "identifier": "timeline.abg81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.euweek.in.quarter",
          "tags": "date eu quarter week",
        },
        "Week (Mon-Sun)/Year (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aa281lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "From - To (Timeline)": Object {
              "identifier": "timeline.aaW81lMifn6q",
              "tags": "",
            },
            "Week #/Year (W1/2010) (Timeline)": Object {
              "identifier": "timeline.aa281lMifn6q",
              "tags": "",
            },
            "Week Starting (Timeline)": Object {
              "identifier": "timeline.aaY81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.euweek",
          "tags": "date eu week",
        },
        "Week (Sun-Sat) (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aaI81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Number US (Timeline)": Object {
              "identifier": "timeline.aaI81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.week.in.year",
          "tags": "date week year",
        },
        "Week (Sun-Sat) of Qtr (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aaO81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Number US (Timeline)": Object {
              "identifier": "timeline.aaO81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.week.in.quarter",
          "tags": "date quarter week",
        },
        "Week (Sun-Sat)/Year (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aaA81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "From - To (Timeline)": Object {
              "identifier": "timeline.aau81lMifn6q",
              "tags": "",
            },
            "Week #/Year (Cont.) (Timeline)": Object {
              "identifier": "timeline.aay81lMifn6q",
              "tags": "",
            },
            "Week #/Year (W1/2010) (Timeline)": Object {
              "identifier": "timeline.aaA81lMifn6q",
              "tags": "",
            },
            "Week Starting (Timeline)": Object {
              "identifier": "timeline.aaw81lMifn6q",
              "tags": "",
            },
            "Wk/Qtr/Year (Cont.) (Timeline)": Object {
              "identifier": "timeline.aaC81lMifn6q",
              "tags": "",
            },
            "Wk/Qtr/Year (Timeline)": Object {
              "identifier": "timeline.aas81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.week",
          "tags": "date week",
        },
        "Year (Timeline)": Object {
          "defaultDisplayForm": Object {
            "identifier": "timeline.aag81lMifn6q",
            "tags": "",
          },
          "displayForms": Object {
            "Year (Timeline)": Object {
              "identifier": "timeline.aag81lMifn6q",
              "tags": "",
            },
          },
          "identifier": "timeline.year",
          "tags": "date year",
        },
      },
      "identifier": "timeline.dataset.dt",
      "tags": "",
    },
  },
  "measures": Object {
    "$ Menu Item Sales": Object {
      "identifier": "fact.salesdetailfact.menuitemsales",
      "tags": "",
    },
    "Cost": Object {
      "identifier": "fact.restaurantcostsfact.cost",
      "tags": "",
    },
    "_Filter Last 4 Quarter": Object {
      "identifier": "aaIHiWZjfWNA",
      "tags": "hidden",
    },
    "_Filter Quarter": Object {
      "identifier": "aaiF4bffe4yn",
      "tags": "hidden",
    },
  },
  "projectId": "test",
  "visualizations": Object {
    "My Renamed Total Sales By Location": Object {
      "identifier": "aaBN5UG3dXu4",
      "tags": "",
    },
    "Total Sales by City": Object {
      "identifier": "abhJpedgcfU2",
      "tags": "",
    },
    "Total Sales by City_1": Object {
      "identifier": "acPWDTL2bJeX",
      "tags": "",
    },
  },
}
`;
